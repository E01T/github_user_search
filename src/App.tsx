import React, { FormEvent } from 'react'

import { MemoizedPageView } from './components/UsersView'
import { MemoizedPagination } from './components/pagination/Pagination'
import { FindUserForm } from './components/FindUserForm'
import { OutputMessage } from './components/OutputMessage'
import { TotalUsers } from './components/TotalUsers'
import { user_type, users_type } from './types/user_types'

const getUser = (user: string, signal: AbortSignal) =>
  fetch(`https://api.github.com/users/${user}`, {
    headers: {
      Accept: 'application/vnd.github.v3+json',
    },
    signal,
  }).then((response) => response.json())

const searchGithubApi = (
  userName: string,
  pageNo: number = 1,
  signal: AbortSignal
) =>
  fetch(
    `https://api.github.com/search/users?q=${userName}&type=users&per_page=10&page=${pageNo}`,
    {
      headers: {
        Accept: 'application/vnd.github.v3+json',
      },
      signal,
    }
  ).then((res) => res.json())

function App() {
  const [users, setUsers] = React.useState<users_type>({
    total_count: 0,
    items: [],
  })

  const [inputValue, setInputValue] = React.useState<string>('') // handle change of input field
  const [userInput, setUserInput] = React.useState<string>('')
  const [pageNo, setPageNo] = React.useState<number>(1)
  const [outputMessage, setOutputMessage] = React.useState<string>('')

  const topRef = React.useRef<HTMLDivElement | null>(null)

  React.useEffect(() => {
    if (userInput === '') return // exit if userInput is empty (initial state)

    const abortCtrl = new AbortController()
    const signal = abortCtrl.signal

    searchGithubApi(userInput, pageNo, signal).then(
      (users) => {
        const total_count: number = users.total_count

        if (total_count === 0) {
          setOutputMessage('No users found')
          return // exit
        }

        const userPromises = users.items.map((user: user_type) => {
          // console.log('user', user)
          return getUser(user.login, signal)
        })

        Promise.all<user_type>(userPromises).then((users) => {
          // console.log('users', users)
          setUsers({ items: users, total_count: total_count })
          topRef?.current?.scrollIntoView({ behavior: 'smooth' })
        })
      },
      (err) => console.error(err)
    )

    return () => abortCtrl.abort()
  }, [userInput, pageNo])

  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    // reset all state
    setUsers({ items: [], total_count: 0 })
    setPageNo(1)
    setInputValue('')
    setOutputMessage('')
    setUserInput(inputValue)
  }

  const handleChange = (e: FormEvent<HTMLInputElement>) => {
    const value = e.currentTarget.value
    setInputValue(value)
  }

  // console.log('users', users)
  return (
    <div ref={topRef} className="container-lg" style={{ margin: '0px auto' }}>
      <div className="container-lg">
        <FindUserForm
          handleSubmit={handleSubmit}
          inputValue={inputValue}
          handleChange={handleChange}
        />
      </div>
      <OutputMessage outputMessage={outputMessage} />
      {users.items.length > 0 ? (
        <div className="container-lg">
          <div className="row d-flex justify-content-center">
            <div className="col-auto">
              <TotalUsers totalUsers={users.total_count} />
            </div>
          </div>
          <div className="row d-flex justify-content-center">
            <MemoizedPageView users={users.items}>
              <MemoizedPagination
                totalCount={users.total_count}
                pageNo={pageNo}
                setPageNo={setPageNo}
              />
            </MemoizedPageView>
          </div>
        </div>
      ) : null}
    </div>
  )
}

export default App
