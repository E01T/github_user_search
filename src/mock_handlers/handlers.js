import { rest } from 'msw'
import MockUsers_10 from '../mock_data/users_10.json'
import MockUsers_15 from '../mock_data/users_15.json'
import SearchAPI_10 from '../mock_data/search_api_10.json'
import SearchAPI_15 from '../mock_data/search_api_15.json'

const __users__ = { foobar: MockUsers_10, 'frontend masters': MockUsers_15 }
const __search__ = { foobar: SearchAPI_10, 'frontend masters': SearchAPI_15 }

const findUser = (usersArray) => (userName) =>
  __users__[usersArray].users.find((user) => user.login === userName)

const findFoobarUser = findUser('foobar')
const findFrontendMastersUser = findUser('frontend masters')

export const handlers = [
  rest.get('https://api.github.com/search/users', async (req, res, ctx) => {
    // const originalResponse = await ctx.fetch(req)
    // const originalResponseData = await originalResponse.json()
    // console.log(originalResponseData.items)

    const query = req.url.searchParams
    const users_q = __search__[query.get('q')]
    if (users_q === undefined)
      return res(ctx.status(404), ctx.json({ total_count: 0, items: [] })) // No users found
    return res(ctx.status(200), ctx.json(users_q))
  }),

  rest.get(`https://api.github.com/users/:user`, async (req, res, ctx) => {
    const foobarUser = findFoobarUser(req.params.user)
    const user = foobarUser
      ? foobarUser
      : findFrontendMastersUser(req.params.user)

    return res(ctx.status(200), ctx.json(user))
  }),
]
