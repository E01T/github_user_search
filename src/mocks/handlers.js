import { rest } from 'msw'
import MockUsers_10 from '../mock_data/users_10.json'
import MockUsers_15 from '../mock_data/users_10.json'
import SearchAPI_10 from '../mock_data/search_api_10.json'
import SearchAPI_15 from '../mock_data/search_api_10.json'
const users_data = [MockUsers_10, MockUsers_15]
const search_data = [SearchAPI_10, SearchAPI_15]

export const handlers = [
  rest.get('https://api.github.com/search/users', async (req, res, ctx) => {
    // const originalResponse = await ctx.fetch(req)
    // const originalResponseData = await originalResponse.json()
    // console.log(originalResponseData.items)
    window.localStorage.setItem('req', req)
    const query = req.url.searchParams
    // console.log('query', query)
    // console.log('rest req', req)
    // console.log('rest req.params', req.params)
    console.log('rest req.url.searchParams', req.url.searchParams)
    const users = search_data[0]
    // console.log('users', users)
    return res(ctx.json(users))
  }),

  rest.get(`https://api.github.com/users/:user`, async (req, res, ctx) => {
    console.log('rest req.params.user', req.params.user)
    return res(
      ctx.json(
        users_data[0].users.find((user) => user.login === req.params.user)
      )
    )
  }),
]
