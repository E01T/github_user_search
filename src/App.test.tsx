import { render, screen, waitFor, cleanup } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { setupServer } from 'msw/node'
import { rest } from 'msw'
import { handlers } from './mock_handlers/handlers'
import App from './App'

// Start the Mock Service Worker
const server = setupServer(...handlers)

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

afterEach(cleanup)

describe('App', () => {
  test('makes 11 http requests when the user clicks the submit btn, searching for uses with the name "foobar"', async () => {
    render(<App />)

    // Why? https://stackoverflow.com/questions/53271193/typeerror-scrollintoview-is-not-a-function
    window.HTMLElement.prototype.scrollIntoView = function () {}
    window.innerWidth = 1200

    const inputElement = screen.getByTestId('search-input')
    const btnElement = screen.getByText(/Submit/i)

    userEvent.type(inputElement, 'foobar')
    inputElement.value = 'foobar'
    userEvent.click(btnElement)

    await waitFor(() => screen.getAllByTestId('card'))
    const users_ = screen.getAllByTestId('card')
    expect(users_.length).toBe(10)
    expect(screen.getByText('Total users: 739')).toBeInTheDocument()
    expect(window.innerWidth).toBe(1200)

    expect(screen.getByText('Previous')).toBeInTheDocument()
    expect(screen.getByText('Next')).toBeInTheDocument()

    expect(screen.getByText('1')).toBeInTheDocument()
    expect(screen.getByText('2')).toBeInTheDocument()
  })

  test('makes 11 http requests when the user clicks the submit btn, searching for uses with the name "frontend masters"', async () => {
    render(<App />)
    const screenWidth = 370
    // Why? https://stackoverflow.com/questions/53271193/typeerror-scrollintoview-is-not-a-function
    window.HTMLElement.prototype.scrollIntoView = function () {}
    window.innerWidth = screenWidth

    const inputElement = screen.getByTestId('search-input')
    const btnElement = screen.getByText(/Submit/i)

    userEvent.type(inputElement, 'frontend masters')
    inputElement.value = 'frontend masters'
    userEvent.click(btnElement)

    await waitFor(() => screen.getAllByTestId('card'))
    const users_ = screen.getAllByTestId('card')
    expect(users_.length).toBe(15)
    expect(screen.getByText('Total users: 15')).toBeInTheDocument()
    expect(window.innerWidth).toBe(screenWidth)

    expect(screen.getByText('Previous')).toBeInTheDocument()
    expect(screen.getByText('Next')).toBeInTheDocument()
  })

  test('makes 1 http requests when the user clicks the submit btn, searching for uses with the name "lX0?qB1&pI6)jS4*"', async () => {
    render(<App />)
    const screenWidth = 1200
    window.innerWidth = screenWidth

    const inputElement = screen.getByTestId('search-input')
    const btnElement = screen.getByText(/Submit/i)

    userEvent.type(inputElement, 'lX0?qB1&pI6)jS4*')
    inputElement.value = 'lX0?qB1&pI6)jS4*'
    userEvent.click(btnElement)

    await waitFor(() => screen.getAllByText('No users found'))
    expect(screen.getByRole('alert')).toBeInTheDocument()
    expect(screen.getByText('No users found')).toBeInTheDocument()
    expect(screen.getByRole('alert').textContent).toMatchSnapshot(
      'No users found'
    )
    expect(window.innerWidth).toBe(screenWidth)
  })

  test('emulates a server error 503 "Service Unavailable"', async () => {
    server.use(
      rest.get('https://api.github.com/search/users', async (req, res, ctx) => {
        return res(
          ctx.status(503),
          ctx.json({
            message: 'Service Unavailable',
          })
        )
      })
    )

    render(<App />)

    const inputElement = screen.getByTestId('search-input')
    const btnElement = screen.getByText(/Submit/i)

    userEvent.type(inputElement, 'whatever')
    inputElement.value = 'whatever'
    userEvent.click(btnElement)

    // sos
    await waitFor(() => screen.getByText('Service Unavailable'), {
      timeout: 2000,
    })

    expect(screen.getByText('Service Unavailable')).toBeInTheDocument()
  })

  test('emulates a generic network error. This triggers the catch block.', async () => {
    server.use(
      rest.get('https://api.github.com/search/users', async (req, res, ctx) => {
        return res.networkError('Internal server error (500)')
      })
    )

    render(<App />)

    const inputElement = screen.getByTestId('search-input')
    const btnElement = screen.getByText(/Submit/i)

    userEvent.type(inputElement, 'any value goes')
    inputElement.value = 'any value goes'
    userEvent.click(btnElement)

    // sos
    await waitFor(() => screen.getByText('Network request failed'), {
      timeout: 2000,
    })

    expect(screen.getByText('Network request failed')).toBeInTheDocument()
  })

  // testing real api
  test.skip('renders input field and submit button', async () => {
    render(<App />)
    // Why? https://stackoverflow.com/questions/53271193/typeerror-scrollintoview-is-not-a-function
    window.HTMLElement.prototype.scrollIntoView = function () {}

    const inputElement = screen.getByTestId('search-input')
    const btnElement = screen.getByText(/Submit/i)

    expect(inputElement).toBeInTheDocument()
    expect(btnElement).toBeInTheDocument()

    userEvent.type(inputElement, 'foobar')
    userEvent.click(btnElement)

    await waitFor(() => screen.findByText('Total users: 739'), {
      timeout: 1000,
    })

    expect(screen.getByText('Total users: 739')).toBeInTheDocument()
    const users_ = screen.getAllByTestId('card')
    expect(users_.length).toBe(10)
  })
})
