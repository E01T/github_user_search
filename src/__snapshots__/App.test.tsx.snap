// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App makes 1 http requests when the user clicks the submit btn, searching for uses with the name "lX0?qB1&pI6)jS4*": No users found 1`] = `"No users found"`;
